### PR TITLE
64 firebase

### PR DESCRIPTION
**Pull request by ibrahim681**
This issue handles Authentication/Authorization using Firebase Admin SDK. The backend is responsible for verifying admins and generating cookies. While the frontend creates new users and calls the backend. This PR only includes backend related code.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@ secretKey.json
 **/pgdata
 output.txt
 **/.DS_Store
+.env
```
Comment (osamaramihafez) [inline .gitignore:6]: https://github.com/Quran-Journey/backend/tree/64-firebase/target
What's the use of this folder? is it generated by firebase?
If it is, let's make sure it's not merged into master by adding it to the gitignore (you'll have to delete it and push to the branch before adding it to gitignore)

Comment (osamaramihafez) [inline .gitignore:6]: https://github.com/Quran-Journey/backend/tree/64-firebase/target
What's the use of this folder? is it generated by firebase?
If it is, let's make sure it's not merged into master by adding it to the gitignore (you'll have to delete it and push to the branch before adding it to gitignore)

File: index.js [hunk 2]
```diff
@@ -38,11 +69,77 @@ app.use("/api", word);
...
+    res.cookie("XSRF-TOKEN", req.csrfToken());
+    next();
+});
+
 app.use(express.static(path.join(__dirname, "/docs")));
 app.route("/").get((req, res) => {
     res.sendFile(path.join(__dirname + "/docs/index.html"));
 });
 
+app.get('/test', (req, res) => {
```
Comment (osamaramihafez) [inline index.js:82]: let's add these endpoints under the routes folder. Can be handled in a separate pr.

File: serviceAccountKey.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (osamaramihafez) [inline serviceAccountKey.json:1]: After MVP, we should really look into storing these securely. Fine for now, same with .env

Comment (osamaramihafez) [inline serviceAccountKey.json:1]: After MVP, we should really look into storing these securely. Fine for now, same with .env

**Review by osamaramihafez — CHANGES_REQUESTED** [review]

File: index.js (hunk 2 above)
Comment (ibrahim681) [reply index.js:82]: any preference for the name of the file?

Comment (ibrahim681) [reply index.js]: I decided testRoutes.js

File: .gitignore (hunk 1 above)
Comment (ibrahim681) [reply .gitignore:6]: deleted this. can't remember what this was for

File: index.js [hunk 4]
```diff
@@ -38,11 +68,71 @@ app.use("/api", word);
...
+//     res.cookie("XSRF-TOKEN", req.csrfToken());
+//     next();
+// });
+
 app.use(express.static(path.join(__dirname, "/docs")));
 app.route("/").get((req, res) => {
     res.sendFile(path.join(__dirname + "/docs/index.html"));
 });
 
+app.get("/home", function (req, res) {
```
Comment (osamaramihafez) [inline index.js:81]: Endpoints for serving html are not necessary

File: index.js [hunk 5]
```diff
@@ -38,11 +68,71 @@ app.use("/api", word);
...
+
+app.get("/login", function (req, res) {
+    res.render("login.html");
+});
+
+app.get("/signup", function (req, res) {
+    res.render("signup.html");
+});
+
+app.get("/profile", function (req, res) {
```
Comment (osamaramihafez) [inline index.js:93]: Take the code from /profile and /sessionLogin and place them in helpers, similar to checkAuth. These endpoints should be referenced in a separate file specific for authentication inside of /routes.

File: index.js [hunk 6]
```diff
@@ -15,20 +19,46 @@ const path = require("path");
...
+
+app.engine("html", require("ejs").renderFile);
 
 app.use(async (req, res, next) => {
     console.log(`\nEndpoint Hit: ${req.method} ${req.originalUrl}\n`);
     next();
 });
 
-
+app.use("/test", checkAuth, testRoutes);
```
Comment (osamaramihafez) [inline index.js]: Rename to /auth

File: index.js [hunk 7]
```diff
@@ -15,20 +19,46 @@ const path = require("path");
 const db = require("./model/db");
 const verseInfo = require("./routes/verseInfo");
 const tafsir = require("./routes/tafsir");
+const testRoutes = require("./routes/testRoutes");
```
Comment (osamaramihafez) [inline index.js:22]: ~/testRoutes~ --> /auth

File: index.js [hunk 8]
```diff
@@ -15,20 +19,46 @@ const path = require("path");
...
 
-var port = 3001;
+const serviceAccount = require("./serviceAccountKey.json");
+
+admin.initializeApp({
+    credential: admin.credential.cert(serviceAccount)
+});
+
+function checkAuth(req, res, next) {
```
Comment (osamaramihafez) [inline index.js:30]: Move to /routes/auth

File: .gitignore (hunk 1 above)
Comment (ibrahim681) [reply .gitignore:6]: everything seems fine without it

File: index.js (hunk 7 above)
Comment (osamaramihafez) [reply index.js:22]: I left the comment on the wrong endpoint. We don't need the /test endpoint at all, we need to move auth endpoints to a separate `/auth` endpoint.

File: routes/testRoutes.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ *  produces:
+ *    - application/json
+ *  responses:
+ *    200:
+ *      description: Hello World!.
+ *    403:
+ *      description: Unauthorized.
+ *
+ */
+router.get("/test", async (request, response) => {
```
Comment (osamaramihafez) [inline routes/testRoutes.js:19]: This file should be for auth (i.e. login, logout... etc) endpoints

**Review by osamaramihafez — CHANGES_REQUESTED** [review]

File: routes/auth.js [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+ *  produces:
+ *    - application/json
+ *  responses:
+ *    200:
+ *      description: redirect.
+ *
+ */
+router.get("/sessionLogout", (req, res) => {
+    res.clearCookie("session");
+    res.redirect("/home");
```
Comment (osamaramihafez) [inline routes/auth.js:111]: Just keep in mind there is no home. We should redirect to the firebase login.

File: index.js [hunk 11]
```diff
@@ -15,20 +19,33 @@ const path = require("path");
...
+var port = process.env.PORT || 3001;
 
 var app = express();
 app.use(cors());
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: true }));
+app.use(cookieParser());
+// app.use(csrfMiddleware);
+
+app.engine("html", require("ejs").renderFile);
```
Comment (osamaramihafez) [inline index.js:42]: Don't think we need this anymore

File: index.js [hunk 12]
```diff
@@ -15,20 +19,33 @@ const path = require("path");
 const db = require("./model/db");
 const verseInfo = require("./routes/verseInfo");
 const tafsir = require("./routes/tafsir");
+const authentication = require("./routes/auth");
+const { checkAuth } = require("./routes/utils");
 
-var port = 3001;
+const serviceAccount = require("./serviceAccountKey.json");
+
+admin.initializeApp({
```
Comment (osamaramihafez) [inline index.js:27]: I'm not sure if this should be here, is it needed here or can it be moved to auth?

**Review by osamaramihafez — APPROVED** [review]

File: index.js (hunk 12 above)
Comment (ibrahim681) [reply index.js:27]: we could, but not sure if that makes the most sense considering the current routes file structure. we can move it to routes/utils.js though.

File: routes/auth.js (hunk 10 above)
Comment (ibrahim681) [reply routes/auth.js:111]: yup will add message to response object instead. great point, thanks!

**Pull request merged**